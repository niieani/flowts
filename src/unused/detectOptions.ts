import * as babel from '@babel/core';
import traverse, { NodePath } from '@babel/traverse';
import { isIdentifier, Program } from '@babel/types';
import { sharedParserPlugins } from '../sharedParserPlugins';

function getReferenceExportName(ref: NodePath): string | undefined {
  let name;
  if (ref.parentPath.isMemberExpression()) {
    if (isIdentifier(ref.parentPath.node.property)) {
      name = ref.parentPath.node.property.name;
    } else {
      ref.parentPath.buildCodeFrameError(
        'unexpected property in member expression'
      );
    }
  } else if (ref.parentPath.isQualifiedTypeIdentifier()) {
    name = ref.parentPath.node.id.name;
  }
  return name;
}

export interface SourceOptions {
  imports: Map<
    string,
    {
      exports: Set<string>;
      default?: Set<string>;
    }
  >;
}

export function detectOptions(source: string, filename: string): SourceOptions {
  const flowAst = babel.parseSync(source, {
    babelrc: false,
    ast: true,
    parserOpts: {
      plugins: ['flow', 'jsx', ...sharedParserPlugins],
    },
    filename,
  });

  if (flowAst === null) {
    throw new Error(
      'babel.parseSync returned null - probably there is some configuration error'
    );
  }

  const imports = new Map<
    string,
    { exports: Set<string>; default?: Set<string> }
  >();

  traverse(flowAst, {
    Program(path: NodePath<Program>) {
      for (const [bindingName, binding] of Object.entries(
        path.scope.bindings
      )) {
        if (binding.kind === 'module') {
          let moduleName: string;
          if (binding.path.parentPath.isImportDeclaration()) {
            moduleName = binding.path.parentPath.node.source.value;
          } else {
            throw path.buildCodeFrameError(
              `Module binding "${bindingName}", expected to be part of import declaration`
            );
          }

          let module = imports.get(moduleName);
          if (!module) {
            module = { exports: new Set() };
            imports.set(moduleName, module);
          }

          if (binding.path.isImportDefaultSpecifier()) {
            const defaultUsages = new Set<string>();
            for (const ref of binding.referencePaths) {
              const name = getReferenceExportName(ref);
              if (name) {
                defaultUsages.add(name);
              }
            }
            module.default = defaultUsages;
          }

          if (binding.path.isImportNamespaceSpecifier()) {
            for (const ref of binding.referencePaths) {
              const name = getReferenceExportName(ref);
              if (name) {
                module.exports.add(name);
              }
            }
          }

          if (binding.path.isImportSpecifier()) {
            module.exports.add(bindingName);
          }
        }
      }
    },
  });
  return { imports };
}
