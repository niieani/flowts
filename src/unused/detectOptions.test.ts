import { detectOptions } from './detectOptions';

describe('detectOptions', () => {
  it('collects module usage info', () => {
    expect(
      detectOptions(
        `
import React from 'react';
import * as N from 'n';
import {render} from 'react-dom';
import type { ComponentType } from 'react';

let a = React.createElement('a');
class A extends N.T {
}

let b: N.B;
`,
        'test.js'
      )
    ).toMatchSnapshot();
  });
});
