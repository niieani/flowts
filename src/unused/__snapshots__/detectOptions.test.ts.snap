// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detectOptions collects module usage info 1`] = `
Object {
  "imports": Map {
    "react" => Object {
      "default": Set {
        "createElement",
      },
      "exports": Set {
        "ComponentType",
      },
    },
    "n" => Object {
      "exports": Set {
        "T",
        "B",
      },
    },
    "react-dom" => Object {
      "exports": Set {
        "render",
      },
    },
  },
}
`;
